{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# This is a script used to calculate the core genome Average Nucleotide Identity of different serotypes of E.coli\n",
    "# Core genomes used in this analysis are Soft core genome defined by Roary\n",
    "# Author: Huifeng Hu(hzauhhf@gmail.com)\n",
    "\n",
    "# Please note: In this analysis, we want to analysis ANI of more than 6,000 genomes of E.coli. \n",
    "# And it is not possible to run Roary with over 6,000 genomes, so we run Roary with 1702 representative genomes and use Blast to find\n",
    "# core genomes in other genomes\n",
    "\n",
    "# THe analysis can be divied into following steps\n",
    "# 1. Get sequence of core gene accoring Roary result(gene_presence_absence.csv)\n",
    "# 2. Use CD-HIT to cluster each gene family and get representative gene sequences\n",
    "# 3. Build a Blastn database with these representative sequences\n",
    "# 4, Use orfs of 6,0000 genomes as query, search against the core gene database\n",
    "# 5. Trim with Blast result and get the best hit result\n",
    "# 6. FastANI was used to calculate pairs of core genomes(Please note parameter \"--fragLen\" must be set according your shortest core gene.)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 1. Get sequence of core gene accoring Roary result(gene_presence_absence.csv)\n",
    "# We use gene id as key and get nucl sequence in gff files;\n",
    "import os\n",
    "gff_dir=r'path/to/gff_files'\n",
    "id2fasta={}\n",
    "for i in os.listdir(gff_dir):\n",
    "    f=open(os.path.join(gff_dir,i),'r')\n",
    "    data=f.read().split('##FASTA')\n",
    "    f.close()\n",
    "    dd={}\n",
    "    for i in data[1].split('>')[1:]:\n",
    "        dd[i.split('\\n')[0]] = i.split('\\n',1)[1].replace('\\n','')\n",
    "    for i in data[0].split('\\n')[:-1]:\n",
    "        if i[0]!= '#':\n",
    "            id2fasta[i.split('\\t')[-1].split(';')[0].split('=')[1]] = dd[i.split('\\t')[0]][int(i.split('\\t')[3])-1:int(i.split('\\t')[4])]\n",
    "import pandas as pd\n",
    "f=pd.read_csv(\"gene_presence_absence.csv\",low_memory=False)\n",
    "ff=f[f['No. isolates']>1619] # Set according roary result\n",
    "d={}\n",
    "for index, row in ff.iterrows():\n",
    "    d[list(row)[0]]=list(row)[14:]\n",
    "out_put_path =r'path/to/your/core/gene/seq'\n",
    "for i in d:\n",
    "    of=open(os.path.join(out_put_path,i),'w')\n",
    "    for j in d[i]:\n",
    "        if type(j)==str:\n",
    "            if '\\t' in j:\n",
    "                for k in j.split('\\t'):\n",
    "                    of.write('>'+k+'\\n'+id2fasta[k]+'\\n')\n",
    "            else:\n",
    "                of.write('>'+j+'\\n'+id2fasta[j]+'\\n')\n",
    "    of.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 2. Use CD-HIT to cluster each gene family and get representative gene sequences\n",
    "# 3. Build a Blastn database with these representative sequences\n",
    "import os\n",
    "for i in os.listdir(r'path/to/your/core/gene/seq'):\n",
    "    if 'py' not in i:\n",
    "        a='cd-hit -i '+i+' -o ../'+i+' -c 0.8 -d 50 -T 8 -M 0'\n",
    "        os.system(a)\n",
    "# 3. Build a Blastn database with these representative sequences        \n",
    "os.system('cat *.representative sequences > core_genomes_DB.fasta')\n",
    "# 4, Use orfs of 6,0000 genomes as query, search against the core gene database\n",
    "os.system('makeblastdb -in core_genomes_DB.fasta -dbtype nucl')\n",
    "os.system('blastn -query *.fna -db core_genomes_DB.fasta -evalue 1e-5 -out *.blast_result -outfmt 6 -num_threads 32')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 5. Trim with Blast result and get the best hit result\n",
    "# We select best hit in Blast result and extract hit range of query sequences as core gene \n",
    "import os\n",
    "def get_best_hit(input_file,output_file):\n",
    "    f=open(input_file,'r')\n",
    "    data=f.read().split('\\n')\n",
    "    f.close()\n",
    "    besthit={}\n",
    "    for i in data[:-1]:\n",
    "        if i.split('\\t')[0] not in besthit:\n",
    "            besthit[i.split('\\t')[0]] = i\n",
    "        else:\n",
    "            pass\n",
    "    of=open(os.path.join(best_hit_path,output_file),'w')\n",
    "    for i in besthit:\n",
    "        of.write(besthit[i]+'\\n')\n",
    "    of.close()\n",
    "\n",
    "blast_result_path=r'path/to/your/blast/result'\n",
    "best_hit_path=r'path/to/your/best_hit'\n",
    "\n",
    "for each_file in os.listdir(blast_result_path):\n",
    "    get_best_hit(os.path.join(blast_result_path,each_file),each_file)\n",
    "\n",
    "nucl_dir = r'path/to/your/nucl_file' \n",
    "core_genome_dir=r'path/to/your/core_genomes'\n",
    "\n",
    "def get_core_genome(x):\n",
    "    d={}\n",
    "    f=open(os.path.join(nucl_dir,x+'.nucl'),'r')\n",
    "    data=f.read().split('>',1)[1].split('\\n>')\n",
    "    f.close()\n",
    "    for i in data:\n",
    "        d[i.split('\\n',1)[0].split(' ')[0]] = i.split('\\n',1)[1].replace('\\n','')\n",
    "    f=open(os.path.join(best_hit_path,x+'_coregenome-hit.out'),'r')\n",
    "    data=f.read().split('\\n')[:-1]\n",
    "    f.close()\n",
    "    of=open(os.path.join(o_dir,x+'.fasta'),'w')\n",
    "    for i in data:\n",
    "        of.write('>'+i.split('\\t')[0]+'\\n'+d[i.split('\\t')[0]]+'\\n')\n",
    "    of.close()\n",
    "        \n",
    "for i in os.listdir(hit_dir):\n",
    "    get_core_genome(i.split('_core')[0])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 6. FastANI was used to calculate pairs of core genomes(Please note parameter \"--fragLen\" must be set according your shortest core gene.)\n",
    "# fastANI were running as followings\n",
    "fastANI -q query_geonomes.fasta -r reference_genomes.fasta -o query:refer -t 'n' --fragLen 100"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
